gr-blocks/python/blocks/matrix_interleaver.py

```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Caliola Engineering, LLC.
+#
+# This is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3, or (at your option)
+# any later version.
+#
+# This software is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this software; see the file COPYING.  If not, write to
+# the Free Software Foundation, Inc., 51 Franklin Street,
+# Boston, MA 02110-1301, USA.
+#
+
+from gnuradio import gr, blocks
+
+class matrix_interleaver(gr.hier_block2):
+    def __init__(self, itemsize, rows=1, cols=1, deint=False):
+        gr.hier_block2.__init__(
+            self, "Matrix Interleaver",
+                gr.io_signature(1, 1, itemsize),
+                gr.io_signature(1, 1, itemsize),
+        )
+
+        self.itemsize = itemsize
+        self.set_rowsandcols(rows, cols, deint)
+
+    def set_rowsandcols(self, rows, cols, deint):
+        self.disconnect_all()
```

Comment:
I'm asking because I honestly don't know the answer: Is it safe to call `self.disconnect_all()` while the flowgraph is running? I usually pause the flowgraph before changing block connections.

Reply:
This block was used in the context of a GRC that had some default settings but could be updated (via variables) prior to a run. The interleave and deinterleave blocks did not allow that at all (no exposed setters if I recall), so this was my solution. The way this block works if the number of connections between the two blocks changes, you'd have to disconnect anyways. 

The purpose was not to allow updates at runtime, but prior to runtime yet after the original flowgraph construction. This may have been specifically to work around GRC behavior more so than GR. I don't think callbacks in GRC (e.g. changing the interleaver block size) prompt a rebuild of the flowgraph, they just call into blocks on the existing flowgraph.

I have never called this callback during runtime, seems like it would be pretty cavalier in this case. Perhaps a disclaimer in the documentation for this use case.

Reply:
Currently, the API should throw an exception or ignore this call during flowgraph execution. If you want to change anything internally before the first call to `start`, I suggest to instantiate a new block.

Reply:
Since this callback shouldn't be called at runtime, is it a good idea to declare it in the GRC YAML file? I haven't tested it, but I suspect that if someone puts a QT GUI variable in GRC for the number of rows, then changes the QT GUI value during runtime, the callback will get called. I think it's better not to let GRC think that the parameters can be changed via callback.

Reply:
I have a multimode transmitter with a set of interleaver options. The interleaver is buried in a hierarchy where I can set a mode with only a single top-level parameter and it propagates down to this block.

Without this capability I would have to have multiple versions of my flow graph with different defaults. This is a huge timesaver for interactive exploration of parameters between runs. I wonder if there's a way to get GRC to disable changes after starting the fg?  It would be a shame to have to workaround the reconfiguration limitations again.